public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -368,6 +378,47 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
         return { status: 'error', message: message };
       });
   }
+
+  async annotationQuery(options: AnnotationQueryRequest<{ expr: string }>): Promise<AnnotationEvent[]> {
+    const annotation = options.annotation;
+    const expr = annotation.expr || '';
+
+    if (!expr) {
+      return [];
+    }
+
+    const target: LokiQuery = {
+      refId: options.annotation.name,
+      expr,
+      live: false,
+    };
+
+    const query: DataQueryRequest<LokiQuery> = {
+      requestId: '',
+      timezone: 'utc',
+      range: {
+        from: dateTime(options.range.from),
+        to: dateTime(options.range.to),
+      },
+      targets: [target],
+    } as any;
+
+    const { data } = await this.runQueries(query);
+    const annotations: AnnotationEvent[] = [];
+    for (const frame of data) {
+      const tags = Object.values(frame.labels);
+      const view = new DataFrameView<{ ts: string; line: string }>(frame);
+      view.forEachRow(row => {
+        annotations.push({
+          time: new Date(row.ts).valueOf(),
+          text: row.line,
+          tags,
```

Comment:
can this just use `frame.labels` directly?  they are not mutated with view/row

Reply:
It needs to be `string[]` instead of keyValue so it is transformed. Or do you mean something else here?

Reply:
My bad -- you are right